### PR TITLE
[aws-sqs] adding support logging to AWS SQS queue

### PR DESCRIPTION
**Pull request by ryandeivert**
There are osquery logger plugins for Kinesis Streams and Firehose, but not for SQS. This (hopefully) adds support for the AWS SQS service.

My cpp is pretty rusty, but I hope this is a good jumping off point for a review. It looks like the dummy class will take care of testing the core functionality for me, but if there is anything I've missed please let me know!!

### Changes

* Adding aws sqs queue logger support via plugin
* Updates toolchain for sqs support (I may have missed something here)
* Documentation updates to includes sqs support and config how-to

cc @guliashvili 
resolves: #4757

### CODE REVIEW

File: osquery/logger/plugins/aws_sqs.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  initAwsSdk();
+  forwarder_ = std::make_shared<SQSLogForwarder>(
+      "aws_sqs", aws_sqs_queue_url, 500);
+  Status s = forwarder_->setUp();
+  if (!s.ok()) {
+    LOG(ERROR) << "Error initializing SQS logger: " << s.getMessage();
+    return s;
+  }
+  Dispatcher::addService(forwarder_);
+  return Status(0, "OK");
```
Comment (obelisk) [inline osquery/logger/plugins/aws_sqs.cpp]: @fmanco What is the decided on style for new Status OK objects now?

Comment (guliashvili) [reply osquery/logger/plugins/aws_sqs.cpp]: Status::success(msg)

File: osquery/logger/plugins/aws_sqs.cpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+Status SQSLogForwarder::internalSetup() {
+  if (FLAGS_aws_sqs_queue_url.empty()) {
+    return Status(1, "Queue URL must be specified with --aws_sqs_queue_url");
+  }
+
+  VLOG(1) << "SQS logging initialized with queue URL: "
+          << FLAGS_aws_sqs_queue_url;
+
+  return Status(0, "OK");
```
Comment (guliashvili) [inline osquery/logger/plugins/aws_sqs.cpp]: Status::success(

File: osquery/logger/plugins/aws_sqs.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+#include <osquery/system.h>
+
+#include "osquery/logger/plugins/aws_sqs.h"
+
+namespace osquery {
+
+REGISTER(SQSLoggerPlugin, "logger", "aws_sqs");
+
+FLAG(uint64,
+     aws_sqs_period,
```
[the remark this answers is not in the document]
Comment (ryandeivert) [reply osquery/logger/plugins/aws_sqs.cpp]: sounds good - will make the change. should the other aws logger plugins be updated to match?

File: osquery/logger/plugins/aws_sqs.cpp (hunk 1 above)
Comment (ryandeivert) [reply osquery/logger/plugins/aws_sqs.cpp]: it looks like msg isn't actually required here and `Status::success()` will suffice?

File: osquery/logger/plugins/aws_sqs.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+void SQSLoggerPlugin::init(const std::string& name,
+                               const std::vector<StatusLogLine>& log) {
+  logStatus(log);
+}
+
+Status SQSLogForwarder::internalSetup() {
+  if (FLAGS_aws_sqs_queue_url.empty()) {
+    return Status(1, "Queue URL must be specified with --aws_sqs_queue_url");
```
Comment (guliashvili) [inline osquery/logger/plugins/aws_sqs.cpp]: Status::failure(

File: osquery/logger/plugins/aws_sqs.h [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#pragma once
+
+#include <chrono>
```
Comment (guliashvili) [inline osquery/logger/plugins/aws_sqs.h]: is it used?

File: osquery/logger/plugins/aws_sqs.h [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+#include <memory>
```
Comment (guliashvili) [inline osquery/logger/plugins/aws_sqs.h]: is it used?

File: osquery/logger/plugins/aws_sqs.h [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+#include <vector>
+
+#include <aws/sqs/SQSClient.h>
+#include <aws/sqs/model/SendMessageBatchRequestEntry.h>
+#include <aws/sqs/model/Message.h>
+
+#include <osquery/core.h>
+#include <osquery/dispatcher.h>
```
Comment (guliashvili) [inline osquery/logger/plugins/aws_sqs.h]: is it used?

File: osquery/logger/plugins/aws_sqs.cpp [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <algorithm>
```
Comment (guliashvili) [inline osquery/logger/plugins/aws_sqs.cpp]: is it used?

File: osquery/logger/plugins/aws_sqs.cpp [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+Status SQSLoggerPlugin::setUp() {
+  initAwsSdk();
+  forwarder_ = std::make_shared<SQSLogForwarder>(
+      "aws_sqs", aws_sqs_queue_url, 500);
+  Status s = forwarder_->setUp();
+  if (!s.ok()) {
+    LOG(ERROR) << "Error initializing SQS logger: " << s.getMessage();
+    return s;
+  }
+  Dispatcher::addService(forwarder_);
```
Comment (guliashvili) [inline osquery/logger/plugins/aws_sqs.cpp]: Returns status lets check

File: osquery/logger/plugins/aws_sqs.cpp [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+     aws_sqs_period,
+     10,
+     "Seconds between flushing logs to SQS (default 10)");
+
+FLAG(string, aws_sqs_queue_url, "", "Name of SQS queue URL for logging")
+
+Status SQSLoggerPlugin::setUp() {
+  initAwsSdk();
+  forwarder_ = std::make_shared<SQSLogForwarder>(
+      "aws_sqs", aws_sqs_queue_url, 500);
```
Comment (guliashvili) [inline osquery/logger/plugins/aws_sqs.cpp]: aws_sqs" is used twice maybe using it as const static is better

File: osquery/logger/plugins/aws_sqs.h [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  size_t getMaxBytesPerBatch() const override;
+  size_t getMaxRetryCount() const override;
+  size_t getInitialRetryDelay() const override;
+  bool appendNewlineSeparators() const override;
+
+  size_t getFailedRecordCount(Outcome& outcome) const override;
+  Result getResult(Outcome& outcome) const override;
+
+ private:
+  FRIEND_TEST(SQSTests, test_send);
```
Comment (guliashvili) [inline osquery/logger/plugins/aws_sqs.h]: SQSTests? looks some debugging leftover. If you have some tests more than welcome to add it.

**Comment by ryandeivert** [comment]
thanks for the feedback @guliashvili - I've added a commit to address your comments, ready for a second round :)

**Comment by guliashvili** [comment]
ok to test

**Comment by guliashvili** [comment]
make format_master to fix code audit

**Comment by ryandeivert** [comment]
ran formatting but unsure what the error is now. locally, this passes:

```
$ make format_check
[+] clang-format (/usr/local/bin/clang-format) version: clang-format version 7.0.0 (tags/google/stable/2018-04-24)
Code passes formatting tests!
```

**Comment by guliashvili** [comment]
make format_master should resolve all the problems by itself. Is not it?

**Comment by guliashvili** [comment]
make format_master does the formatting on the diff between or branch and the master. Your master does not look  to be clean ( = real master). Probably that's why it misses to format some things.

**Comment by ryandeivert** [comment]
ah thank - it's probably outdated, I'll update and try again :)

**Comment by ryandeivert** [comment]
hey @guliashvili I'm a bit confused as to what is failing here.. none of the failures I'm seeing seem to be related to any of my changes. could you please point me in the right direction to resolve?

**Comment by guliashvili** [comment]
@ryandeivert 
Build is multithreaded so error might not be in the end.
For example in this [build](https://jenkins.osquery.io/job/osqueryPullRequestLinux/8320/TargetSystem=aws-ubuntu-14.04/console)  
related problem looks to be

> error[ 45%] Building CXX object osquery/tables/CMakeFiles/osquery_system_posix_tables.dir/system/posix/yum_sources.cpp.o
> : 'aws/sqs/model/SendMessageBatchRequest.h' file not found
> #include <aws/sqs/model/SendMessageBatchRequest.h>
>          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by ryandeivert** [comment]
thanks @guliashvili !! It appears this is happening because the brew bottle for the `aws-sdk-cpp` dependencies does not have the required sqs libs.

example:
```bash
$ aws s3 cp s3://osquery-packages/bottles/aws-sdk-cpp-1.4.55_1.x86_64_linux.bottle.tar.gz ./
$ tar xvf aws-sdk-cpp-1.4.55_1.x86_64_linux.bottle.tar.gz 
$ tree -L 3 aws-sdk-cpp
aws-sdk-cpp
└── 1.4.55_1
    ├── INSTALL_RECEIPT.json
    ├── LICENSE
    ├── LICENSE.txt
    ├── NOTICE.txt
    ├── README.md
    ├── include
    │   └── aws
    └── lib
        ├── cmake
        ├── libaws-cpp-sdk-core.a
        ├── libaws-cpp-sdk-ec2.a
        ├── libaws-cpp-sdk-firehose.a
        ├── libaws-cpp-sdk-kinesis.a
        ├── libaws-cpp-sdk-sts.a
        └── libtesting-resources.a

5 directories, 11 files
```

From what I've read in the docs, this seems like it's something I need someone else to update on my behalf? Are you someone that could assist with that? Thanks again!

**Comment by guliashvili** [comment]
In an effort to keep our open PR list clean and easily actionable we're closing stale pull requests. If you plan to work on this PR in the future simply re-open the PR when you have updates and we're happy to follow up. Check our [contributing](https://github.com/facebook/osquery/blob/master/CONTRIBUTING.md) guidelines or comment below if you have questions. Thank you.

**Pull request closed without merging**